x/ccv/provider/keeper/genesis.go

```diff
@@ -22,39 +23,146 @@ func (k Keeper) InitGenesis(ctx sdk.Context, genState *types.GenesisState) {
 		}
 	}
 
+	k.SetValidatorSetUpdateId(ctx, genState.ValsetUpdateId)
+	for _, v2h := range genState.ValsetUpdateIdToHeight {
+		k.SetValsetUpdateBlockHeight(ctx, v2h.ValsetUpdateId, v2h.Height)
+	}
+
+	for _, cccp := range genState.ConsumerAdditionProposals {
+		// prevent implicit memory aliasing
+		cccp := cccp
+		if err := k.SetPendingConsumerAdditionProp(ctx, &cccp); err != nil {
+			panic(fmt.Errorf("pending create consumer chain proposal could not be persisted: %w", err))
+		}
+	}
+	for _, sccp := range genState.ConsumerRemovalProposals {
+		k.SetPendingConsumerRemovalProp(ctx, sccp.ChainId, sccp.StopTime)
+	}
+	for _, ubdOp := range genState.UnbondingOps {
+		if err := k.SetUnbondingOp(ctx, ubdOp); err != nil {
+			panic(fmt.Errorf("unbonding op could not be persisted: %w", err))
+		}
+	}
+
+	if genState.MatureUnbondingOps != nil {
+		if err := k.AppendMaturedUnbondingOps(ctx, genState.MatureUnbondingOps.Ids); err != nil {
+			panic(err)
+		}
+	}
+
 	// Set initial state for each consumer chain
-	for _, cc := range genState.ConsumerStates {
-		k.SetChainToChannel(ctx, cc.ChainId, cc.ChannelId)
-		k.SetChannelToChain(ctx, cc.ChannelId, cc.ChainId)
+	for _, cs := range genState.ConsumerStates {
+		chainID := cs.ChainId
+		k.SetConsumerClientId(ctx, chainID, cs.ClientId)
+		if err := k.SetConsumerGenesis(ctx, chainID, cs.ConsumerGenesis); err != nil {
+			panic(fmt.Errorf("consumer chain genesis could not be persisted: %w", err))
+		}
+		if cs.LockUnbondingOnTimeout {
+			k.SetLockUnbondingOnTimeout(ctx, chainID)
+		}
+		// check if the CCV channel was established
+		if cs.ChannelId != "" {
+			k.SetChannelToChain(ctx, cs.ChannelId, chainID)
+			k.SetChainToChannel(ctx, chainID, cs.ChannelId)
+			k.SetInitChainHeight(ctx, chainID, cs.InitialHeight)
+
+			k.SetSlashAcks(ctx, cs.ChainId, cs.SlashDowntimeAck)
+			for _, ubdOpIndex := range cs.UnbondingOpsIndex {
+				k.SetUnbondingOpIndex(ctx, chainID, ubdOpIndex.ValsetUpdateId, ubdOpIndex.UnbondingOpIndex)
+			}
+		} else {
+			for _, vsc := range cs.PendingValsetChanges {
+				k.AppendPendingVSC(ctx, chainID, vsc)
+			}
+		}
 	}
 
 	k.SetParams(ctx, genState.Params)
 }
 
 func (k Keeper) ExportGenesis(ctx sdk.Context) *types.GenesisState {
-	store := ctx.KVStore(k.storeKey)
-	iterator := sdk.KVStorePrefixIterator(store, []byte{types.ChannelToChainBytePrefix})
-	defer iterator.Close()
-
-	if !iterator.Valid() {
-		return types.DefaultGenesisState()
-	}
-
 	var consumerStates []types.ConsumerState
+	// export states for each consumer chains
+	k.IterateConsumerChains(ctx, func(ctx sdk.Context, chainID, clientID string) bool {
+		gen, found := k.GetConsumerGenesis(ctx, chainID)
+		if !found {
```

Comment:
Do we want to have the consumer genesis stored on the provider for ever, or should we allow the consumer genesis to be deleted once the consumer bootstraps?